ft: increase crr metrics expiry window to 24hrs

philipyoo opened the pull request:
Changes in this PR:
- Increase metrics expiry, but keep throughput to 15 minute
  averages.
- Add helper method `_getMaxUptime` to find # of intervals
- Update tests to reflect the extra intervals fetched from
  Redis/StatsModel
- Remove `OBJECT_MONITORING_EXPIRY` and use `EXPIRY` instead
  as values are now same
- Use single instance of `StatsModel`
- Remove extra interval in `StatsModel` expiry field. Not
  needed anymore as throughput uses a 15 minute window and
  the extra interval for it will be available by default
- Use explicit variable names when data is fetched from
  `StatsClient`

bert-e commented:
# Hello philipyoo,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get
information on this process.





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @RachedBenMustapha

* @bennettbuchanan

philipyoo commented:
@bert-e wait

bennettbuchanan commented on lib/backbeat/Metrics.js:
```diff
@@ -4,7 +4,8 @@ const errors = require('../../lib/errors');
-const EXPIRY = 900; // 15 minutes
+const EXPIRY = 86400; // 24 hours
+const THROUGHPUT_EXPIRY = 900; // 15 minutes
 const OBJECT_MONITORING_EXPIRY = 86400; // 24 hours.
```
I think we can go ahead and get rid of `OBJECT_MONITORING_EXPIRY` since it was added only because it was different from the original expiry time.

bennettbuchanan commented on lib/backbeat/Metrics.js:
```diff
@@ -14,8 +15,7 @@ class Metrics {
         // the immediate older data for average throughput calculation
-        this._statsClient = new StatsModel(this._redisClient, INTERVAL,
-            (EXPIRY + INTERVAL));
+        this._statsClient = new StatsModel(this._redisClient, INTERVAL, EXPIRY);
```
How come we don't want to pass in `EXPIRY + INTERVAL` as the third argument to the `StatsModel` constructor?

philipyoo replied:
the extra interval was used for throughput to get a more accurate average. Keeping throughput to still use a 15 minute range, so we no longer need this extra interval 👍

bennettbuchanan replied:
Oh, I see, since it will be included anyway in the new expiry time frame. 👍

bennettbuchanan commented on lib/backbeat/Metrics.js:
```diff
@@ -96,6 +96,16 @@ class Metrics {
+     * @return {number} uptime of server up to expiry time
+     */
+    _getMaxUptime(expiry) {
+        const timeSinceStart = (Date.now() - this._internalStart) / 1000;
```
Minor: Renaming `timeSinceStart` to `secondsSinceStart` might help with readability here.

bennettbuchanan commented on lib/backbeat/Metrics.js:
```diff
@@ -336,20 +336,18 @@ class Metrics {
                 // all intervals apply, including 4th interval
                 const lastInterval =
-                    this._statsClient._normalizeTimestamp(new Date(now));
+                    this._statsClient._normalizeTimestamp(now);
```
👍 Nice improvement along the way!

bennettbuchanan commented on tests/functional/backbeat/Metrics.js:
```diff
@@ -57,15 +61,16 @@ describe('Metrics class', () => {
                 completions: {
                     description: 'Number of completed replication operations' +
                         ' (count) and number of bytes transferred (size) in ' +
-                        'the last 900 seconds',
```
Let's be sure to update the docs to reflect this change. For example, [here](https://github.com/scality/backbeat/blob/development/8.0/docs/metrics.md#_backbeatapimetricscrrlocation-namecompletions) and other references to the default window of 15 minutes. That should help prevent a perception that the routes aren't working as intended.

bennettbuchanan commented on lib/backbeat/Metrics.js:
```diff
@@ -181,10 +190,10 @@ class Metrics {
                     description: 'Number of completed replication operations ' +
                         '(count) and number of bytes transferred (size) in ' +
-                        `the last ${Math.floor(timeDisplay)} seconds`,
+                        `the last ${Math.floor(uptime)} seconds`,
```
If `uptime` is `1` here, then `Math.floor` will make the value `0` which seems inaccurate.

philipyoo replied:
I think `Math.floor` here will still leave it at 1, `uptime` will be guaranteed at least 1

bennettbuchanan replied:
Wow, apparently I don't know how `Math.floor` works! It does, however, look like `uptime` could be a decimal < 1. It's definitely a minor case though, since probably no requests will come in before one second has elapsed and it only affects the description 😸.

philipyoo replied:
Ah you're right! I pushed a change to guarantee minimum value of 1 in the helper method `_getMaxUptime`

bennettbuchanan reviewed (approved):
Awesome speedy turnaround on this change! Overall, LGTM.

rahulreddy reviewed (changes requested):
@philipyoo This is just to make our maintenance easier. Can you separate the commits, one for changing the expiry and two for the refactor you are doing here and add context in the commit body.

philipyoo committed:
ft: ZENKO-925 increase crr metrics expiry to 24hrs

Changes reflected in this commit:
- Increase metrics expiry, but keep throughput to 15 minute
  averages.
- Add helper method `_getMaxUptime` to find # of intervals
- Update tests to reflect the extra intervals fetched from
  Redis/StatsModel

philipyoo committed:
rf: use single StatsModel, use explicit var names

Changes in this commit:
- Remove `OBJECT_MONITORING_EXPIRY` and use `EXPIRY` instead
  as values are now same
- Use single instance of `StatsModel`
- Remove extra interval in `StatsModel` expiry field. Not
  needed anymore as throughput uses a 15 minute window and
  the extra interval for it will be available by default
- Use explicit variable names when data is fetched from
  `StatsClient`

jonathan-gramain commented on lib/backbeat/Metrics.js:
```diff
@@ -96,6 +97,18 @@ class Metrics {
+    _getMaxUptime(expiry) {
+        let secondsSinceStart = (Date.now() - this._internalStart) / 1000;
+        // allow only a minimum value of 1 for uptime
+        if (secondsSinceStart < 1) secondsSinceStart = 1;
```
I think the style guidelines advocate to always use curly braces and put the conditional statement(s) in this block on separate lines.

jonathan-gramain commented:
BTW, is this change relevant for S3 connector? Not sure but since the plan is to have retry feature available in S3 connector it could be worth targeting development/7.4.

philipyoo commented:
@jonathan-gramain Ya, I think this should also apply for s3 connector, but there is quite a big difference how metrics looks in 7.4. For 7.4, these functions only exists on backbeat, so I'll need to create separate PRs for these changes on backbeat. Will create a jira ticket for tracking this. Also, I have a few other things I need to port for metrics

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.0`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of Release Engineering will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of Release Engineering now.

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.0`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.4`



Please check the status of the associated issue ZENKO-925.

Goodbye philipyoo.
